src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs

```diff
@@ -22,329 +21,185 @@ namespace System.Threading
+        private static partial void GetOrCreateLockObject(ObjectHandleOnStack obj, ObjectHandleOnStack lockObj);
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static void Enter_Slowpath(object obj)
-        {
-            Enter_Slowpath(ObjectHandleOnStack.Create(ref obj));
-        }
+        #endregion
 
-        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Monitor_TryEnter_Slowpath")]
-        private static partial int TryEnter_Slowpath(ObjectHandleOnStack obj, int timeout);
+        #region Object->Condition mapping
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static bool TryEnter_Slowpath(object obj)
+        private static readonly ConditionalWeakTable<object, Condition> s_conditionTable = [];
+        private static readonly Func<object, Condition> s_createCondition = (o) => new Condition(GetLockObject(o));
+
+        private static Condition GetCondition(object obj)
         {
-            if (TryEnter_Slowpath(ObjectHandleOnStack.Create(ref obj), 0) != 0)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
+            Debug.Assert(
+                obj is not Condition,
+                "Do not use Monitor.Pulse or Wait on a Condition instance; use the methods on Condition instead.");
+            return s_conditionTable.GetOrAdd(obj, s_createCondition);
         }
+        #endregion
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static bool TryEnter_Slowpath(object obj, int timeout)
+        #region Public Enter/Exit methods
+        public static void Enter(object obj)
         {
-            ArgumentOutOfRangeException.ThrowIfLessThan(timeout, -1, null);
+            ObjectHeader.HeaderLockResult result = ObjectHeader.TryAcquireThinLock(obj);
+            if (result == ObjectHeader.HeaderLockResult.Success)
+                return;
 
-            if (TryEnter_Slowpath(ObjectHandleOnStack.Create(ref obj), timeout) != 0)
-            {
-                return true;
-            }
-            else
-            {
-                return false;
-            }
+            GetLockObject(obj).Enter();
         }
 
-        // Use a ref bool instead of out to ensure that unverifiable code must
-        // initialize this value to something.  If we used out, the value
-        // could be uninitialized if we threw an exception in our prolog.
-        // The JIT should inline this method to allow check of lockTaken argument to be optimized out
-        // in the typical case. Note that the method has to be transparent for inlining to be allowed by the VM.
         public static void Enter(object obj, ref bool lockTaken)
         {
+            // we are inlining lockTaken check as the check is likely be optimized away
             if (lockTaken)
-                ThrowLockTakenException();
```

Comment:
Keep `ThrowLockTakenException` throw helper to make this more likely to be inlined and prevent eating JIT inlining budgets?